From 88687c9939c5d744b1f7224d77943099b659ab17 Mon Sep 17 00:00:00 2001
From: Yuri Nesterov <yuriy.nesterov@unikie.com>
Date: Wed, 9 Apr 2025 22:24:52 +0300
Subject: [PATCH] usb-host: enable autoscan for bus+addr to survive host
 suspend/resume

Currently, there is a special case for USB devices added using the
hostbus= and hostaddr= properties to avoid adding them to the hotplug
watchlist, since the address changes every time the device is plugged
in. However, when the host system goes into suspend and then resumes,
those devices are disconnected from the guest with the "no device"
error but their address stays the same. Enabling autoscan and adding
them to the watchlist allows them to be reconnected to guest after host
suspend/resume.

Signed-off-by: Yuri Nesterov <yuriy.nesterov@unikie.com>
---
 hw/usb/host-libusb.c | 10 ++++++----
 1 file changed, 6 insertions(+), 4 deletions(-)

diff --git a/hw/usb/host-libusb.c b/hw/usb/host-libusb.c
index 691bc881fb..5c1e740bd7 100644
--- a/hw/usb/host-libusb.c
+++ b/hw/usb/host-libusb.c
@@ -1242,12 +1242,14 @@ static void usb_host_realize(USBDevice *udev, Error **errp)
                        s->match.bus_num, s->match.addr);
             return;
         }
-    } else {
-        s->needs_autoscan = true;
-        QTAILQ_INSERT_TAIL(&hostdevs, s, next);
-        usb_host_auto_check(NULL);
     }
 
+    /* Enabling autoscan for all match types allows all attached devices,
+     * including those added by bus+addr, to survive host suspend/resume */
+    s->needs_autoscan = true;
+    QTAILQ_INSERT_TAIL(&hostdevs, s, next);
+    usb_host_auto_check(NULL);
+
     s->exit.notify = usb_host_exit_notifier;
     qemu_add_exit_notifier(&s->exit);
 }
-- 
2.43.0
